File: frontend/passes/liveness.go
package passes

import (
	"github.com/mattcarp12/maml/frontend/mir"
	"github.com/mattcarp12/maml/frontend/types"
)

type LivenessResult struct {
	LiveIn  map[mir.BlockID]map[string]bool
	LiveOut map[mir.BlockID]map[string]bool
	Aliases map[string]string
}

type BlockStatementLiveness struct {
	LiveIn      []map[string]bool
	LiveOut     []map[string]bool
	TermLiveIn  map[string]bool
	TermLiveOut map[string]bool
}

// Helper to clone a set so we don't accidentally mutate shared maps
func cloneSet(s map[string]bool) map[string]bool {
	clone := make(map[string]bool, len(s))
	for k, v := range s {
		clone[k] = v
	}
	return clone
}

func AnalyzeLiveness(g *mir.Graph, locals map[string]types.Type) *LivenessResult {
	res := &LivenessResult{
		LiveIn:  make(map[mir.BlockID]map[string]bool),
		LiveOut: make(map[mir.BlockID]map[string]bool),
		Aliases: buildAliasMap(g, locals),
	}

	for id := range g.Blocks {
		res.LiveIn[id] = make(map[string]bool)
		res.LiveOut[id] = make(map[string]bool)
	}

	blockUses := make(map[mir.BlockID]map[string]bool)
	blockDefs := make(map[mir.BlockID]map[string]bool)
	for _, block := range g.Blocks {
		useSet, defSet := computeBlockUseDef(block, res.Aliases)
		blockUses[block.ID] = useSet
		blockDefs[block.ID] = defSet
	}

	changed := true
	for changed {
		changed = false

		maxID := mir.BlockID(0)
		for id := range g.Blocks {
			if id > maxID {
				maxID = id
			}
		}

		for id := maxID; id >= 0; id-- {
			block, exists := g.Blocks[id]
			if !exists {
				continue
			}

			// 1. LiveOut = Union of LiveIn of all successors
			succs := getSuccessors(block)
			for _, succID := range succs {
				for v := range res.LiveIn[succID] {
					if !res.LiveOut[id][v] {
						res.LiveOut[id][v] = true
						changed = true
					}
				}
			}

			useSet := blockUses[id]
			defSet := blockDefs[id]

			// 2. LiveIn = Use U (LiveOut - Def)
			// Track modifications accurately to verify fixed-point completion
			for v := range useSet {
				if !res.LiveIn[id][v] {
					res.LiveIn[id][v] = true
					changed = true
				}
			}
			for v := range res.LiveOut[id] {
				if !defSet[v] {
					if !res.LiveIn[id][v] {
						res.LiveIn[id][v] = true
						changed = true
					}
				}
			}
		}
	}

	return res
}

func computeBlockUseDef(block *mir.BasicBlock, aliases map[string]string) (map[string]bool, map[string]bool) {
	useSet := make(map[string]bool)
	defSet := make(map[string]bool)

	// Helper to add a use and its resolved alias
	addUse := func(name string) {
		useSet[name] = true
		if root := resolveAlias(name, aliases); root != name {
			useSet[root] = true
		}
	}

	// --- STEP 1: Process Terminator Uses FIRST ---
	termUses := getTerminatorUses(block.Terminator)
	for _, u := range termUses {
		addUse(u)
	}

	// --- STEP 2: Traverse Statements BACKWARD (Bottom to Top) ---
	for i := len(block.Statements) - 1; i >= 0; i-- {
		inst := block.Statements[i]

		// Use the single source of truth for instruction use/defs
		uses, defs := getInstUseDef(inst)

		// A Def kills any upward Use path inside this block
		for _, d := range defs {
			defSet[d] = true
			delete(useSet, d)
		}

		// Add uses and their aliases
		for _, u := range uses {
			addUse(u)
		}
	}

	return useSet, defSet
}

// AnalyzeStatementLiveness computes the exact liveness across every instruction
// in a single block. This is required for Non-Lexical Lifetime (NLL) borrow checking.
func AnalyzeStatementLiveness(block *mir.BasicBlock, blockLiveOut map[string]bool, aliases map[string]string) *BlockStatementLiveness {
	res := &BlockStatementLiveness{
		LiveIn:      make([]map[string]bool, len(block.Statements)),
		LiveOut:     make([]map[string]bool, len(block.Statements)),
		TermLiveIn:  make(map[string]bool),
		TermLiveOut: cloneSet(blockLiveOut),
	}

	// 1. Start with the block's global LiveOut
	currentLive := cloneSet(blockLiveOut)

	// 2. Process the Terminator
	termUses := getTerminatorUses(block.Terminator)
	for _, u := range termUses {
		currentLive[u] = true
		if root := resolveAlias(u, aliases); root != u {
			currentLive[root] = true
		}
	}
	res.TermLiveIn = cloneSet(currentLive)

	// 3. Walk statements backward
	for i := len(block.Statements) - 1; i >= 0; i-- {
		inst := block.Statements[i]

		// The LiveOut of this instruction is the current live set
		res.LiveOut[i] = cloneSet(currentLive)

		uses, defs := getInstUseDef(inst)

		// Equation: LiveIn = (LiveOut - Defs) U Uses

		// First, kill the defs
		for _, d := range defs {
			delete(currentLive, d)
		}

		// Then, gen the uses
		for _, u := range uses {
			currentLive[u] = true
			if root := resolveAlias(u, aliases); root != u {
				currentLive[root] = true
			}
		}

		// The new set is the LiveIn for this instruction
		res.LiveIn[i] = cloneSet(currentLive)
	}

	return res
}

func getTerminatorUses(term mir.Terminator) []string {
	var uses []string
	addUse := func(op mir.Value) {
		if reg, ok := op.(*mir.Register); ok && reg != nil {
			uses = append(uses, reg.Name)
		}
	}

	switch t := term.(type) {
	case *mir.BranchTerminator:
		addUse(t.Condition)
	case *mir.ReturnTerminator:
		if t.Value != nil {
			addUse(t.Value)
		}
	case *mir.CoroSuspendTerminator:
		// Suspend terminators themselves don't typically use local registers directly
	}
	return uses
}

func getInstUseDef(inst mir.Instruction) (uses []string, defs []string) {
	addUseVal := func(op mir.Value) {
		if reg, ok := op.(*mir.Register); ok && reg != nil {
			uses = append(uses, reg.Name)
		}
	}
	addUseName := func(name string) {
		if name != "" && name != "_" {
			uses = append(uses, name)
		}
	}
	addDef := func(name string) {
		if name != "" && name != "_" {
			defs = append(defs, name)
		}
	}

	switch i := inst.(type) {
	case *mir.AssignInst:
		addDef(i.Dst)
		addUseVal(i.RValue)
	case *mir.CopyInst:
		addDef(i.Dst)
		addUseName(i.Src)
	case *mir.MoveInst:
		addDef(i.Dst)
		addUseName(i.Src)
	case *mir.BorrowInst:
		addDef(i.Dst)
		addUseName(i.Src)
	case *mir.CastInst:
		addDef(i.Dst)
		addUseVal(i.Src)
	case *mir.BinaryOpInst:
		addDef(i.Dst)
		addUseVal(i.Left)
		addUseVal(i.Right)
	case *mir.UnaryOpInst:
		addDef(i.Dst)
		addUseVal(i.Operand)
	case *mir.BitcastPtrInst:
		addDef(i.Dst)
		addUseVal(i.Src)
	case *mir.FieldAddrInst:
		addDef(i.Dst)
		addUseVal(i.Object)
	case *mir.LoadPtrInst:
		addDef(i.Dst)
		addUseVal(i.Ptr)
	case *mir.StoreInst:
		addUseVal(i.DstPtr)
		addUseVal(i.Value)
	case *mir.IndexAddrInst:
		addDef(i.Dst)
		addUseVal(i.Source)
		addUseVal(i.Index)
	case *mir.CallInst:
		if i.Dst != "" && i.Dst != "_" {
			addDef(i.Dst)
		}
		addUseVal(i.Function)
		for _, arg := range i.Arguments {
			addUseVal(arg)
		}
	case *mir.KeepAliveInst:
		addUseName(i.Src)
	}

	return uses, defs
}

func buildAliasMap(g *mir.Graph, locals map[string]types.Type) map[string]string {
	aliases := make(map[string]string)

	isView := func(name string) bool {
		t, ok := locals[name]
		if !ok {
			return false
		}
		_, ok = t.(*types.ViewType)
		return ok
	}

	for _, block := range g.Blocks {
		for _, inst := range block.Statements {
			switch i := inst.(type) {
			case *mir.BorrowInst:
				aliases[i.Dst] = i.Src
			case *mir.FieldAddrInst:
				if reg, ok := i.Object.(*mir.Register); ok {
					aliases[i.Dst] = reg.Name
				}
			case *mir.IndexAddrInst:
				if reg, ok := i.Source.(*mir.Register); ok {
					aliases[i.Dst] = reg.Name
				}
			case *mir.AddressOfInst:
				aliases[i.Dst] = i.Src
			case *mir.LoadPtrInst:
				if reg, ok := i.Ptr.(*mir.Register); ok {
					if root := resolveAlias(reg.Name, aliases); root != reg.Name {
						aliases[i.Dst] = root
					}
				}
			case *mir.BitcastPtrInst:
				if reg, ok := i.Src.(*mir.Register); ok {
					aliases[i.Dst] = reg.Name
				}
			case *mir.CallInst:
				if reg, ok := i.Function.(*mir.Register); ok {
					if reg.Name == "maml_vec_get" || reg.Name == "maml_map_get" {
						if len(i.Arguments) > 0 {
							if argReg, ok := i.Arguments[0].(*mir.Register); ok {
								aliases[i.Dst] = argReg.Name
							}
						}
					}
				}
			case *mir.CoroPromisePtrInst:
				if reg, ok := i.Handle.(*mir.Register); ok {
					aliases[i.Dst] = reg.Name
				}
			case *mir.StoreInst:
				if srcReg, isReg := i.Value.(*mir.Register); isReg {
					if dstPtrReg, isPtrReg := i.DstPtr.(*mir.Register); isPtrReg {
						dstRoot := resolveAlias(dstPtrReg.Name, aliases)
						if isView(dstRoot) {
							srcRoot := resolveAlias(srcReg.Name, aliases)
							if srcRoot != dstRoot {
								aliases[dstRoot] = srcRoot
							}
						}
					}
				}
			}

			// --- NEW: Track alias transfers for Move, Copy, and Assignment ---
			if mv, ok := inst.(*mir.MoveInst); ok && isView(mv.Dst) {
				if root := resolveAlias(mv.Src, aliases); root != mv.Src {
					aliases[mv.Dst] = root
				}
			}
			
			if cp, ok := inst.(*mir.CopyInst); ok && isView(cp.Dst) {
				if root := resolveAlias(cp.Src, aliases); root != cp.Src {
					aliases[cp.Dst] = root
				}
			}
			
			if as, ok := inst.(*mir.AssignInst); ok && isView(as.Dst) {
				if reg, ok := as.RValue.(*mir.Register); ok {
					if root := resolveAlias(reg.Name, aliases); root != reg.Name {
						aliases[as.Dst] = root
					}
				}
			}
		}
	}
	return aliases
}

func resolveAlias(name string, aliases map[string]string) string {
	if parent, exists := aliases[name]; exists && parent != name {
		return resolveAlias(parent, aliases)
	}
	return name
}

File: passes/src/liveness.cpp
#include "liveness.h"

namespace maml::passes {

template <class... Ts> struct overloaded : Ts... {
    using Ts::operator()...;
};
template <class... Ts> overloaded(Ts...) -> overloaded<Ts...>;

// =============================================================================
// Top-Level Solvers
// =============================================================================

LivenessResult LivenessAnalyzer::analyzeLiveness(
    const mir::Graph* g, const std::unordered_map<SymID, const types::Type*>& locals)
{
    LivenessResult res;
    res.aliases = buildAliasMap(g, locals);

    for (const auto& [id, block] : g->blocks) {
        res.liveIn[id] = {};
        res.liveOut[id] = {};
    }

    std::unordered_map<mir::BlockID, std::unordered_set<SymID>> blockUses;
    std::unordered_map<mir::BlockID, std::unordered_set<SymID>> blockDefs;

    for (const auto& [id, block] : g->blocks) {
        auto [u, d] = computeBlockUseDef(block.get(), res.aliases);
        blockUses[id] = std::move(u);
        blockDefs[id] = std::move(d);
    }

    bool changed = true;
    mir::BlockID maxID = 0;
    for (const auto& [id, block] : g->blocks) {
        if (id > maxID)
            maxID = id;
    }

    // Fixed-point backward dataflow[cite: 3]
    while (changed) {
        changed = false;

        for (mir::BlockID id = maxID; id >= 0; id--) {
            auto it = g->blocks.find(id);
            if (it == g->blocks.end())
                continue;

            const mir::BasicBlock* block = it->second.get();

            // 1. LiveOut = Union of LiveIn of all successors[cite: 3]
            std::vector<mir::BlockID> succs = getSuccessors(block);
            for (mir::BlockID succID : succs) {
                for (SymID v : res.liveIn[succID]) {
                    if (res.liveOut[id].insert(v).second) {
                        changed = true;
                    }
                }
            }

            const auto& useSet = blockUses[id];
            const auto& defSet = blockDefs[id];

            // 2. LiveIn = Use U (LiveOut - Def)[cite: 3]
            for (SymID v : useSet) {
                if (res.liveIn[id].insert(v).second) {
                    changed = true;
                }
            }

            for (SymID v : res.liveOut[id]) {
                if (defSet.find(v) == defSet.end()) {
                    if (res.liveIn[id].insert(v).second) {
                        changed = true;
                    }
                }
            }
        }
    }

    return res;
}

BlockStatementLiveness LivenessAnalyzer::analyzeStatementLiveness(const mir::BasicBlock* block,
    const std::unordered_set<SymID>& blockLiveOut, const std::unordered_map<SymID, SymID>& aliases)
{
    BlockStatementLiveness res;
    size_t numStmts = block->statements.size();
    res.liveIn.resize(numStmts);
    res.liveOut.resize(numStmts);
    res.termLiveOut = blockLiveOut;

    std::unordered_set<SymID> currentLive = blockLiveOut;

    // Process Terminator[cite: 3]
    std::vector<SymID> termUses = getTerminatorUses(block->terminator);
    for (SymID u : termUses) {
        currentLive.insert(u);
        SymID root = resolveAlias(u, aliases);
        if (root != u)
            currentLive.insert(root);
    }
    res.termLiveIn = currentLive;

    // Traverse statements backward[cite: 3]
    for (int i = static_cast<int>(numStmts) - 1; i >= 0; i--) {
        const auto& inst = block->statements[i];
        res.liveOut[i] = currentLive;

        std::vector<SymID> uses, defs;
        getInstUseDef(inst, uses, defs);

        // LiveIn = (LiveOut - Defs) U Uses[cite: 3]
        for (SymID d : defs) {
            currentLive.erase(d);
        }

        for (SymID u : uses) {
            currentLive.insert(u);
            SymID root = resolveAlias(u, aliases);
            if (root != u)
                currentLive.insert(root);
        }

        res.liveIn[i] = currentLive;
    }

    return res;
}

// =============================================================================
// Block-Level Extractors
// =============================================================================

std::pair<std::unordered_set<SymID>, std::unordered_set<SymID>>
LivenessAnalyzer::computeBlockUseDef(
    const mir::BasicBlock* block, const std::unordered_map<SymID, SymID>& aliases)
{
    std::unordered_set<SymID> useSet;
    std::unordered_set<SymID> defSet;

    auto addUse = [&](SymID name) {
        if (name == NoSymbol)
            return;
        useSet.insert(name);
        SymID root = resolveAlias(name, aliases);
        if (root != name)
            useSet.insert(root);
    };

    std::vector<SymID> termUses = getTerminatorUses(block->terminator);
    for (SymID u : termUses)
        addUse(u);

    for (int i = static_cast<int>(block->statements.size()) - 1; i >= 0; i--) {
        std::vector<SymID> uses, defs;
        getInstUseDef(block->statements[i], uses, defs);

        for (SymID d : defs) {
            if (d == NoSymbol)
                continue;
            defSet.insert(d);
            useSet.erase(d);
        }

        for (SymID u : uses)
            addUse(u);
    }

    return { useSet, defSet };
}

// =============================================================================
// Instruction Extractors
// =============================================================================

std::vector<mir::BlockID> LivenessAnalyzer::getSuccessors(const mir::BasicBlock* block)
{
    std::vector<mir::BlockID> succs;
    std::visit(overloaded {
                   [&](std::monostate) {},
                   [&](const mir::JumpTerminator& t) { succs.push_back(t.target); },
                   [&](const mir::BranchTerminator& t) {
                       succs.push_back(t.trueTarget);
                       succs.push_back(t.falseTarget);
                   },
                   [&](const mir::CoroSuspendTerminator& t) {
                       succs.push_back(t.resumeBlock);
                       succs.push_back(t.cleanupBlock);
                   },
                   [&](auto&) {} // Others have no intra-function successors
               },
        block->terminator);
    return succs;
}

std::vector<SymID> LivenessAnalyzer::getTerminatorUses(const mir::Terminator& term)
{
    std::vector<SymID> uses;
    auto addUse = [&](const mir::Value& val) {
        if (auto* reg = std::get_if<mir::Register>(&val))
            uses.push_back(reg->name);
    };

    std::visit(
        overloaded { [&](std::monostate) {},
            [&](const mir::BranchTerminator& t) { addUse(t.condition); },
            [&](const mir::ReturnTerminator& t) { addUse(t.value); },
            [&](const mir::CoroFinalSuspendTerminator& t) { addUse(t.value); }, [&](auto&) {} },
        term);
    return uses;
}

void LivenessAnalyzer::getInstUseDef(
    const mir::Instruction& inst, std::vector<SymID>& uses, std::vector<SymID>& defs)
{
    auto addUseVal = [&](const mir::Value& val) {
        if (auto* reg = std::get_if<mir::Register>(&val)) {
            if (reg->name != NoSymbol)
                uses.push_back(reg->name);
        }
    };
    auto addUseName = [&](SymID name) {
        if (name != NoSymbol)
            uses.push_back(name);
    };
    auto addDef = [&](SymID name) {
        if (name != NoSymbol)
            defs.push_back(name);
    };

    std::visit(
        overloaded { [&](std::monostate) {}, [&](const mir::AllocaInst& i) { addDef(i.dst); },
            [&](const mir::AssignInst& i) {
                addDef(i.dst);
                addUseVal(i.rValue);
            },
            [&](const mir::CopyInst& i) {
                addDef(i.dst);
                addUseName(i.src);
            },
            [&](const mir::MoveInst& i) {
                addDef(i.dst);
                addUseName(i.src);
            },
            [&](const mir::BorrowInst& i) {
                addDef(i.dst);
                addUseName(i.src);
            },
            [&](const mir::CastInst& i) {
                addDef(i.dst);
                addUseVal(i.src);
            },
            [&](const mir::BinaryOpInst& i) {
                addDef(i.dst);
                addUseVal(i.left);
                addUseVal(i.right);
            },
            [&](const mir::UnaryOpInst& i) {
                addDef(i.dst);
                addUseVal(i.operand);
            },
            [&](const mir::BitcastPtrInst& i) {
                addDef(i.dst);
                addUseVal(i.src);
            },
            [&](const mir::FieldAddrInst& i) {
                addDef(i.dst);
                addUseVal(i.object);
            },
            [&](const mir::LoadPtrInst& i) {
                addDef(i.dst);
                addUseVal(i.ptr);
            },
            [&](const mir::StoreInst& i) {
                addUseVal(i.dstPtr);
                addUseVal(i.value);
            },
            [&](const mir::IndexAddrInst& i) {
                addDef(i.dst);
                addUseVal(i.source);
                addUseVal(i.index);
            },
            [&](const mir::CallInst& i) {
                if (i.dst != NoSymbol)
                    addDef(i.dst);
                addUseVal(i.function);
                for (const auto& arg : i.arguments)
                    addUseVal(arg);
            },
            [&](const mir::KeepAliveInst& i) { addUseName(i.src); }, [&](auto&) {} },
        inst);
}

// =============================================================================
// Alias Graph Tracking
// =============================================================================

std::unordered_map<SymID, SymID> LivenessAnalyzer::buildAliasMap(
    const mir::Graph* g, const std::unordered_map<SymID, const types::Type*>& locals)
{
    std::unordered_map<SymID, SymID> aliases;

    auto isView = [&](SymID name) {
        auto it = locals.find(name);
        if (it == locals.end() || it->second == nullptr)
            return false;
        return it->second->kind == types::TypeKind::View;
    };

    for (const auto& [id, block] : g->blocks) {
        for (const auto& inst : block->statements) {
            std::visit(
                overloaded { [&](std::monostate) {},
                    [&](const mir::BorrowInst& i) { aliases[i.dst] = i.src; },
                    [&](const mir::AddressOfInst& i) { aliases[i.dst] = i.src; },
                    [&](const mir::FieldAddrInst& i) {
                        if (auto* reg = std::get_if<mir::Register>(&i.object))
                            aliases[i.dst] = reg->name;
                    },
                    [&](const mir::IndexAddrInst& i) {
                        if (auto* reg = std::get_if<mir::Register>(&i.source))
                            aliases[i.dst] = reg->name;
                    },
                    [&](const mir::LoadPtrInst& i) {
                        if (auto* reg = std::get_if<mir::Register>(&i.ptr)) {
                            SymID root = resolveAlias(reg->name, aliases);
                            if (root != reg->name)
                                aliases[i.dst] = root;
                        }
                    },
                    [&](const mir::BitcastPtrInst& i) {
                        if (auto* reg = std::get_if<mir::Register>(&i.src))
                            aliases[i.dst] = reg->name;
                    },
                    [&](const mir::CallInst& i) {
                        if (auto* reg = std::get_if<mir::Register>(&i.function)) {
                            std::string_view name = sym_.resolve(reg->name);
                            if ((name == "maml_vec_get" || name == "maml_map_get")
                                && !i.arguments.empty()) {
                                if (auto* argReg = std::get_if<mir::Register>(&i.arguments[0])) {
                                    aliases[i.dst] = argReg->name;
                                }
                            }
                        }
                    },
                    [&](const mir::CoroPromisePtrInst& i) {
                        if (auto* reg = std::get_if<mir::Register>(&i.handle))
                            aliases[i.dst] = reg->name;
                    },
                    [&](const mir::StoreInst& i) {
                        if (auto* srcReg = std::get_if<mir::Register>(&i.value)) {
                            if (auto* dstReg = std::get_if<mir::Register>(&i.dstPtr)) {
                                SymID dstRoot = resolveAlias(dstReg->name, aliases);
                                if (isView(dstRoot)) {
                                    SymID srcRoot = resolveAlias(srcReg->name, aliases);
                                    if (srcRoot != dstRoot)
                                        aliases[dstRoot] = srcRoot;
                                }
                            }
                        }
                    },
                    // Track alias transfers for Views[cite: 3]
                    [&](const mir::MoveInst& i) {
                        if (isView(i.dst)) {
                            SymID root = resolveAlias(i.src, aliases);
                            if (root != i.src)
                                aliases[i.dst] = root;
                        }
                    },
                    [&](const mir::CopyInst& i) {
                        if (isView(i.dst)) {
                            SymID root = resolveAlias(i.src, aliases);
                            if (root != i.src)
                                aliases[i.dst] = root;
                        }
                    },
                    [&](const mir::AssignInst& i) {
                        if (isView(i.dst)) {
                            if (auto* reg = std::get_if<mir::Register>(&i.rValue)) {
                                SymID root = resolveAlias(reg->name, aliases);
                                if (root != reg->name)
                                    aliases[i.dst] = root;
                            }
                        }
                    },
                    [&](auto&) {} },
                inst);
        }
    }
    return aliases;
}

SymID LivenessAnalyzer::resolveAlias(SymID name, const std::unordered_map<SymID, SymID>& aliases)
{
    std::unordered_set<SymID> visited;
    SymID current = name;

    while (true) {
        visited.insert(current);
        auto it = aliases.find(current);

        // Found root[cite: 3]
        if (it == aliases.end() || it->second == current) {
            return current;
        }

        // Cycle detected, break safely[cite: 3]
        if (visited.find(it->second) != visited.end()) {
            return current;
        }

        current = it->second;
    }
}

} // namespace maml::passes